packages/docusaurus-mdx-loader/src/remark/toc/index.ts

```diff
@@ -93,66 +93,158 @@ const plugin: Plugin = function plugin(
     const {toString} = await import('mdast-util-to-string');
     const {visit} = await import('unist-util-visit');
 
-    const headings: TOCItem[] = [];
+    const partialComponentToHeadingsName: {[key: string]: string} =
+      Object.create(null);
+
+    const headings: (TOCItem | NestedTOC)[] = [];
+    const imports: ImportDeclaration[] = [];
 
-    visit(root, 'heading', (child: Heading) => {
-      const value = toString(child);
+    function visitHeading(node: Heading) {
+      const value = toString(node);
 
       // depth:1 headings are titles and not included in the TOC
-      if (!value || child.depth < 2) {
+      if (!value || node.depth < 2) {
         return;
       }
 
       headings.push({
-        value: toValue(child, toString),
-        id: child.data!.id!,
-        level: child.depth,
+        value: toValue(node, toString),
+        id: node.data!.id!,
+        level: node.depth,
       });
+    }
+
+    function visitMdxjsEsm(node: MdxjsEsm) {
+      if (!node.data?.estree) {
+        return;
+      }
+
+      for (const potentialImportDeclaration of node.data.estree.body) {
+        if (potentialImportDeclaration.type !== 'ImportDeclaration') {
+          continue;
+        }
+
+        const importPath = potentialImportDeclaration.source.value as string;
+        const isMdxImport = /\.mdx?$/.test(importPath);
+        if (!isMdxImport) {
+          continue;
+        }
+
+        const componentName = potentialImportDeclaration.specifiers.find(
+          (o: Node) => o.type === 'ImportDefaultSpecifier',
+        )?.local.name;
+
+        if (!componentName) {
+          continue;
+        }
+        const {length} = Object.keys(partialComponentToHeadingsName);
+        const exportAsName = `${name}${length}`;
+        partialComponentToHeadingsName[componentName] = exportAsName;
+
+        const specifier: ImportSpecifier = {
+          type: 'ImportSpecifier',
+          imported: {type: 'Identifier', name},
+          local: {type: 'Identifier', name: exportAsName},
+        };
+
+        imports.push({
+          type: 'ImportDeclaration',
+          specifiers: [specifier],
+          source: potentialImportDeclaration.source,
+        });
+        potentialImportDeclaration.specifiers.push(specifier);
+      }
+    }
+
+    function visitMdxJsxFlowElement(node: MdxJsxFlowElement) {
+      const nodeName = node.name;
+      if (!nodeName) {
+        return;
+      }
+      const headingsName = partialComponentToHeadingsName[nodeName];
+      if (headingsName) {
+        headings.push({
+          nested: true,
+          name: headingsName,
+        });
+      }
+    }
+
+    visit(root, ['heading', 'mdxjsEsm', 'mdxJsxFlowElement'], (child) => {
```

Comment:
```suggestion
    visit(root, (child) => {
```

This test is useless if you are already testing.
And the types are better if you don’t use arrays. TypeScript doesn’t understand the types of arrays.

Reply:
Removed the array, but can't figure out how to remove the casts. Checking for `child.type` has no effect on the type of `child` inside the `if` statement.

Did I not fully understand you?

Reply:
> Checking for child.type has no effect on the type of child inside the if statement.

It should, if `root` is typed as a `Root`.
It looks like the types here don’t work because `Plugin` is `any`.
Typically, plugins have a type such as `Plugin<[], Root>` would type it.
You should alternatively be able to type `root` on L92 (`root: Root`). Or on L89: `Transformer<Root>`.

Reply:
Can confirm I still see casting is needed after the change

<img width="1009" alt="CleanShot 2024-01-18 at 16 23 09@2x" src="https://github.com/facebook/docusaurus/assets/749374/50d20656-6561-4e61-85aa-67ee3915585b">